### PR TITLE
Add GradDiv to Autograd 

### PR DESCRIPTION
**Pull request by kamathhrishi**
## Description
This PR adds GradDiv functionality. 
Also includes small changes such as 
1) Adding parameters function to Conv2d to allow parameters to be passed to the optimizer
2) Added MPCTensor reciprocal function with test 

## How has this been tested?
Added unit tests for the same. 

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: src/sympc/api.py [hunk 1]
```diff
@@ -247,6 +247,10 @@
         "sympc.tensor.replicatedshare_tensor.ReplicatedSharedTensor.view",
         "sympc.tensor.replicatedshare_tensor.ReplicatedSharedTensor",
     ),
+    (
+        "sympc.tensor.replicatedshare_tensor.ReplicatedSharedTensor.reshape",
```
Comment (gmuraru) [inline src/sympc/api.py:251]: Q: I don't think you need this, right?

Comment (kamathhrishi) [reply src/sympc/api.py:251]: Not for this pr. I was just adding in general, because it will be used in future. I found the use when reshaping outputs for neural networks and thought I should add it.

Comment (gmuraru) [reply src/sympc/api.py:251]: Could you remove it from here and we can add it when we need it?

File: src/sympc/grads/grad_functions.py [hunk 2]
```diff
@@ -351,6 +351,44 @@ def backward(ctx: Dict[str, Any], grad: MPCTensor) -> MPCTensor:
...
+        Args:
+            ctx (Dict[str, Any]): Context used to retrieve the information for the backward pass
+            grad (MPCTensor): The gradient that came from the child nodes
+
+        Returns:
+            (x_grad, y_grad) (Tuple[MPCTensor]): The gradients passed to the X and Y nodes.
+        """
+        x, y = ctx["x"], ctx["y"]
+
+        grad_x = grad * (x / (y * x))
```
Comment (gmuraru) [inline src/sympc/grads/grad_functions.py:386]: Q: Why does it look like this?
```
dx x/y = 1/y
```
So for this, you will have `grad / y `.

File: src/sympc/grads/grad_functions.py [hunk 3]
```diff
@@ -351,6 +351,44 @@ def backward(ctx: Dict[str, Any], grad: MPCTensor) -> MPCTensor:
...
+        grad_y = ((-1) * (x)) / ((y ** 2) * grad)
```
Comment (gmuraru) [inline src/sympc/grads/grad_functions.py:387]: MAYBE: Why not save in the forward the `x / y` and then here multiply by `(-1 * x * grad_cache * grad)`

File: tests/sympc/grad/grad_functions_test.py [hunk 4]
```diff
@@ -814,3 +815,53 @@ def test_grad_maxpool_2d_backward(get_clients, kernel_size, stride, padding) ->
...
+    y_secret = torch.tensor([9.0, 8.0])
+    y = MPCTensor(secret=y_secret, session=session, requires_grad=True)
+
+    grad_mpc = MPCTensor(secret=grad, session=session, requires_grad=True)
+
+    ctx = {"x": x, "y": y}
+
+    grad_x, grad_y = GradDiv.backward(ctx, grad_mpc)
+
+    expected_grad_x = grad * (x_secret / (y_secret * x_secret))
```
Comment (gmuraru) [inline tests/sympc/grad/grad_functions_test.py:860]: Can you test here with torch? (like do division and on the resulting tensor do the backward pass and it should have approximately the same result)

File: tests/sympc/grad/grad_functions_test.py [hunk 5]
```diff
@@ -814,3 +815,53 @@ def test_grad_maxpool_2d_backward(get_clients, kernel_size, stride, padding) ->
...
+    expected_grad_y = ((-1) * (x_secret)) / ((y_secret ** 2) * grad)
+
+    res_x = grad_x.reconstruct()
+    res_y = grad_y.reconstruct()
+
+    assert np.allclose(res_x, expected_grad_x, rtol=1e-1)
```
Comment (gmuraru) [inline tests/sympc/grad/grad_functions_test.py:866]: Q: Can we have here `1e-2`

**Review by gmuraru — CHANGES_REQUESTED** [review]
See comments

File: src/sympc/grads/grad_functions.py (hunk 2 above)
Comment (aanurraj) [reply src/sympc/grads/grad_functions.py:386]: I dont understand this diff :/

File: src/sympc/module/nn/conv.py [hunk 6]
```diff
@@ -140,11 +141,25 @@ def share_state_dict(
...
         self.weight = MPCTensor(secret=weight, session=self.session, shape=shape)
+        self._parameters = OrderedDict({"weight": self.weight})
 
         if bias is not None:
             self.bias = MPCTensor(
                 secret=bias, session=self.session, shape=(self.out_channels,)
             )
+            self._parameters["bias"] = self.bias
+
+    def parameters(self, recurse: bool = False) -> MPCTensor:
```
Comment (aanurraj) [inline src/sympc/module/nn/conv.py:152]: Q: Does this solves the None issue in step while training?

Comment (kamathhrishi) [reply src/sympc/module/nn/conv.py:152]: No, this doesn't. The whole GradDiv PR solves that issue. Basically, MSELoss has a division operation. GradDiv hasn't defined some gradients after the division was None.

File: tests/sympc/grad/grad_functions_test.py (hunk 4 above)
Comment (kamathhrishi) [reply tests/sympc/grad/grad_functions_test.py:860]: Yeah sure. But, I won't just do (x/y) and take gradient. More like (x/y).sum() and take gradient because (x/y) is not a scalar. Anyway, sum has gradient 1.

File: tests/sympc/grad/grad_functions_test.py (hunk 5 above)
Comment (kamathhrishi) [reply tests/sympc/grad/grad_functions_test.py:866]: The tests were failing because there were small differences and the numbers are also small. The output in plaintext was: _tensor([-0.0268, -0.0234])_ and the result after reconstruction:  _tensor([-0.0247, -0.0234])_. Instead of keeping it this way I will change the values being passed (x,y, grad) so that gradients are small enough and we can use 1e-2.

File: src/sympc/grads/grad_functions.py (hunk 2 above)
Comment (kamathhrishi) [reply src/sympc/grads/grad_functions.py:386]: Good question. I have no idea :P but I am figuring it out. 

I know grad(x/y) wrt to x is (1/x). 
When I first made it (1/y) and compared results with PyTorch for the same, they were very different. But, with some random experimentation I found that (x/(x*y) matches Pytorch's grad for division.

File: src/sympc/api.py (hunk 1 above)
Comment (kamathhrishi) [reply src/sympc/api.py:251]: Alright.

File: tests/sympc/grad/grad_functions_test.py (hunk 4 above)
Comment (kamathhrishi) [reply tests/sympc/grad/grad_functions_test.py]: Nevermind.

File: src/sympc/grads/grad_functions.py (hunk 2 above)
Comment (kamathhrishi) [reply src/sympc/grads/grad_functions.py:389]: Nevermind this. I will correct it. I made a mistake.

File: src/sympc/grads/grad_functions.py [hunk 7]
```diff
@@ -351,6 +351,48 @@ def backward(ctx: Dict[str, Any], grad: MPCTensor) -> MPCTensor:
...
+            ctx (Dict[str, Any]): Context used to retrieve the information for the backward pass
+            grad (MPCTensor): The gradient that came from the child nodes
+
+        Returns:
+            (x_grad, y_grad) (Tuple[MPCTensor]): The gradients passed to the X and Y nodes.
+        """
+        x, y = ctx["x"], ctx["y"]
+
+        grad_x = grad * (1 / y)
+        grad_y = (-1) * grad * (ctx["result"] / y)
```
Comment (rasswanth-s) [inline src/sympc/grads/grad_functions.py:390]: MAYBE : grad_y = (-1)*grad_x *ctx["result"]
We could save a division operation,I think division is compute intensive due to approximations.

File: src/sympc/grads/grad_functions.py (hunk 3 above)
Comment (rasswanth-s) [reply src/sympc/grads/grad_functions.py:391]: Q:We should remove the previous method?

File: src/sympc/grads/grad_functions.py [hunk 8]
```diff
@@ -135,6 +135,12 @@ def forward(ctx: Dict[str, Any], x: MPCTensor, y: Any) -> MPCTensor:
         Returns:
             x + y (MPCTensor): The result of the addition
         """
+        if not hasattr(x, "shape"):
```
Comment (gmuraru) [inline src/sympc/grads/grad_functions.py:138]: Insert meme with "we do not do that here"

Comment (gmuraru) [reply src/sympc/grads/grad_functions.py:138]: Can't you leave it only for the tensor case at the moment? this should be treated better

File: src/sympc/module/nn/conv.py [hunk 9]
```diff
@@ -57,6 +57,7 @@ def __init__(self, session: Session) -> None:
         self.padding = 0
         self.dilation = 1
         self.groups = 1
+        self._parameters = None
```
Comment (gmuraru) [inline src/sympc/module/nn/conv.py:60]: Q: Could you add those to slots and change the slots to a tuple from the list?

File: src/sympc/grads/grad_functions.py (hunk 8 above)
Comment (kamathhrishi) [reply src/sympc/grads/grad_functions.py:138]: Nope, it can't be only for tensor case. The reciprocal function uses addition with an integer. So this throws an error saying _"Integer has no attribute shape"_

File: src/sympc/module/nn/conv.py (hunk 9 above)
Comment (kamathhrishi) [reply src/sympc/module/nn/conv.py:60]: Yeah sure.

File: src/sympc/grads/grad_functions.py (hunk 8 above)
Comment (kamathhrishi) [reply src/sympc/grads/grad_functions.py:138]: I can just cast the integers in reciprocal as torch tensors and add it instead.

Comment (kamathhrishi) [reply src/sympc/grads/grad_functions.py]: Done

File: src/sympc/module/nn/conv.py (hunk 9 above)
Comment (kamathhrishi) [reply src/sympc/module/nn/conv.py:62]: Done

File: src/sympc/approximations/reciprocal.py [hunk 10]
```diff
@@ -31,7 +34,7 @@ def reciprocal(self, method: str = "NR", nr_iters: int = 10):
 
     if method == "nr":
         new_self = modulus(self)
-        result = 3 * exp(0.5 - new_self) + 0.003
+        result = 3 * exp(0.5 - new_self) + torch.tensor(0.003)
```
Comment (gmuraru) [inline src/sympc/approximations/reciprocal.py:37]: Q: Can't you remove the tensor from here and do it in the "GradDiv":
```

if isinstance(x, (list, int, float)):
   x = torch.tensor(x)
```

File: tests/sympc/grad/grad_functions_test.py [hunk 11]
```diff
@@ -814,3 +815,55 @@ def test_grad_maxpool_2d_backward(get_clients, kernel_size, stride, padding) ->
...
+    session.autograd_active = True
+    SessionManager.setup_mpc(session)
+
+    x_secret = torch.tensor([1.0, 2.0, 3.0, 4.0], requires_grad=True)
+    x = MPCTensor(secret=x_secret, session=session, requires_grad=True)
+
+    y_secret = torch.tensor([2.0, 3.0, 4.0, 5.0], requires_grad=True)
+    y = MPCTensor(secret=y_secret, session=session, requires_grad=True)
+
+    z = (x_secret / y_secret).sum()
```
Comment (gmuraru) [inline tests/sympc/grad/grad_functions_test.py:852]: OUGHTTO: Can you remove the `sum` from here?
For the moment this test should check only the `div` functionality (you can pass in the gradient when doing `backward`)

File: tests/sympc/tensor/mpc_tensor_test.py [hunk 12]
```diff
@@ -787,3 +787,17 @@ def test_session_ring_xor(get_clients, security, bit) -> None:
     result = operator.xor(x, b)
     expected_res = secret_x ^ secret_b
     assert (result.reconstruct(decode=False) == expected_res).all()
+
+
+def test_mpctensor_reciprocal(get_clients):
```
Comment (gmuraru) [inline tests/sympc/tensor/mpc_tensor_test.py:792]: MAYBE: Could you leave only `test_reciprocal` no need to specify mpc tensor since it is in the `mpc tensor file`

File: src/sympc/approximations/reciprocal.py (hunk 10 above)
Comment (kamathhrishi) [reply src/sympc/approximations/reciprocal.py:37]: In gradsum you mean?

File: tests/sympc/grad/grad_functions_test.py (hunk 11 above)
Comment (kamathhrishi) [reply tests/sympc/grad/grad_functions_test.py]: Done

File: tests/sympc/tensor/mpc_tensor_test.py (hunk 12 above)
Comment (kamathhrishi) [reply tests/sympc/tensor/mpc_tensor_test.py]: Done

File: src/sympc/approximations/reciprocal.py (hunk 10 above)
Comment (kamathhrishi) [reply src/sympc/approximations/reciprocal.py]: Done

File: src/sympc/grads/grad_functions.py [hunk 13]
```diff
@@ -135,6 +135,12 @@ def forward(ctx: Dict[str, Any], x: MPCTensor, y: Any) -> MPCTensor:
         Returns:
             x + y (MPCTensor): The result of the addition
         """
+        if isinstance(x, (list, int, float)):
```
Comment (gmuraru) [inline src/sympc/grads/grad_functions.py:138]: MAYBE: I think you should remove this because it reached this point because x was an MPCTensor, right?

File: src/sympc/grads/grad_functions.py [hunk 14]
```diff
@@ -243,6 +252,12 @@ def forward(ctx: Dict[str, Any], x: MPCTensor, y: Any) -> MPCTensor:
         Returns:
             x - y (MPCTensor): The result of the substraction
         """
+        if isinstance(x, (list, int, float)):
```
Comment (gmuraru) [inline src/sympc/grads/grad_functions.py:255]: Also here

Comment (kamathhrishi) [reply src/sympc/grads/grad_functions.py:255]: Right will remove. Dint realize.

File: src/sympc/grads/grad_functions.py [hunk 15]
```diff
@@ -281,9 +290,10 @@ def forward(ctx: Dict[str, Any], x: MPCTensor, y: Any) -> MPCTensor:
         Returns:
             x * y (MPCTensor): The result of the multiplication
         """
-        if not hasattr(x, "shape"):
+        if isinstance(x, (list, int, float)):
```
Comment (gmuraru) [inline src/sympc/grads/grad_functions.py:293]: You should check only `y`

File: src/sympc/grads/grad_functions.py [hunk 16]
```diff
@@ -351,6 +361,50 @@ def backward(ctx: Dict[str, Any], grad: MPCTensor) -> MPCTensor:
...
+        Args:
+            ctx (Dict[str, Any]): Context used to retrieve the information for the backward pass
+            grad (MPCTensor): The gradient that came from the child nodes
+
+        Returns:
+            (x_grad, y_grad) (Tuple[MPCTensor]): The gradients passed to the X and Y nodes.
+        """
+        y = ctx["y"]
+
+        grad_x = grad * (1 / y)
```
Comment (gmuraru) [inline src/sympc/grads/grad_functions.py:402]: Q: Can't you do `grad / y`?

File: tests/sympc/grad/grad_functions_test.py [hunk 17]
```diff
@@ -814,3 +815,55 @@ def test_grad_maxpool_2d_backward(get_clients, kernel_size, stride, padding) ->
...
+    x = MPCTensor(secret=x_secret, session=session, requires_grad=True)
+
+    y_secret = torch.tensor([2.0, 3.0, 4.0, 5.0], requires_grad=True)
+    y = MPCTensor(secret=y_secret, session=session, requires_grad=True)
+
+    z = x_secret / y_secret
+    z.backward(torch.tensor([1, 1, 1, 1]))
+
+    grad = torch.tensor([1, 1, 1, 1])
+    grad_mpc = MPCTensor(secret=grad, session=session, requires_grad=True)
```
Comment (gmuraru) [inline tests/sympc/grad/grad_functions_test.py:856]: :+1: :+1: :+1:

File: src/sympc/grads/grad_functions.py [hunk 18]
```diff
@@ -173,6 +176,9 @@ def forward(ctx: Dict[str, Any], x: MPCTensor) -> MPCTensor:
         Returns:
             sum(x) (MPCTensor): The summation of all the elements from the input
         """
+        if isinstance(x, (list, int, float)):
```
Comment (rasswanth-s) [inline src/sympc/grads/grad_functions.py:179]: As george mentioned in previous comments, should we omit checking for x here?

File: src/sympc/grads/grad_functions.py [hunk 19]
```diff
@@ -351,6 +361,50 @@ def backward(ctx: Dict[str, Any], grad: MPCTensor) -> MPCTensor:
...
+
+        Args:
+            ctx (Dict[str, Any]): Context used to save information needed in the backward pass
+            x (MPCTensor): 1st operand for the division operation
+            y (Any): 2nd operand for the divison operation
+
+        Returns:
+            x / y (MPCTensor): The result of the division
+        """
+        if isinstance(x, (list, int, float)):
```
Comment (rasswanth-s) [inline src/sympc/grads/grad_functions.py:379]: Q: same question ,Should we check for x here?

File: src/sympc/tensor/mpc_tensor.py [hunk 20]
```diff
@@ -576,6 +576,18 @@ def rmatmul(self, y: torch.Tensor) -> "MPCTensor":
 
         return result
 
+    def rtruediv(self, y: Union["MPCTensor", torch.Tensor, float, int]) -> "MPCTensor":
```
Comment (rasswanth-s) [inline src/sympc/tensor/mpc_tensor.py:579]: Since it is an `rtruediv`, I think y does not include `MPCTensor`

File: src/sympc/tensor/mpc_tensor.py [hunk 21]
```diff
@@ -576,6 +576,18 @@ def rmatmul(self, y: torch.Tensor) -> "MPCTensor":
...
+        """Apply recriprocal of MPCTensor.
+
+        Args:
+            y (Union["MPCTensor", torch.Tensor, float, int]): Numerator.
```
Comment (rasswanth-s) [inline src/sympc/tensor/mpc_tensor.py:583]: Here also

File: tests/sympc/tensor/mpc_tensor_test.py [hunk 22]
```diff
@@ -787,3 +787,17 @@ def test_session_ring_xor(get_clients, security, bit) -> None:
...
     expected_res = secret_x ^ secret_b
     assert (result.reconstruct(decode=False) == expected_res).all()
+
+
+def test_reciprocal(get_clients):
+    clients = get_clients(2)
+    session = Session(parties=clients)
+    SessionManager.setup_mpc(session)
+
+    x_secret = torch.Tensor([1, 2, 3, 4])
```
Comment (rasswanth-s) [inline tests/sympc/tensor/mpc_tensor_test.py:797]: MAYBE: Includes also float values?

**Review by rasswanth-s — APPROVED** [review]
LGTM! Some Comments.

File: tests/sympc/grad/grad_functions_test.py (hunk 17 above)
Comment (kamathhrishi) [reply tests/sympc/grad/grad_functions_test.py:856]: This is fine right?

**Review by gmuraru — APPROVED** [review]
LGTM!
